### ISSUE TITLE
Hoisting

### ISSUE THREAD

**Issue by billychen0894** [billychen0894/blog#27]
# What is Hoisting?
In this article [Execution Context](https://github.com/billychen0894/blog/issues/26), we touched a bit on `Hoisting`. But what is exactly Hoisting? It basically means that you can execute the variables/functions prior to the line of code that's actually declared or invoked.

> Many people misunderstood that hoisting is what automatically push the variables/functions declaration up to the beginning of the code. But this is not 100% accurate. The JavaScript engine doesn't physically hoist the code up in the beginning of the script.

That is, `Hoisting` is all about how `memory is allocated` for variable/function declarations in the `Creation phase` of an `Execution Context`.

# Variable Hoisting
In many other programming languages, if we access the variable before it has been declared, it should give an error. However, it’s different in JavaScript.

In JavaScript, the execution context is divided into two phases - `Creation` and `Execution` phase.

In the creation phase, the memory gets allocated for the variables and always initialized with a special value called `undefined`. (The meaning of `undefined` is, a variable has been declared, but it has no value assigned yet) The creation phase declares the variable by allocating memory for it and marks it as `undefined`. This phenomenon is called `Variable Hoisting` in JavaScript.

```jsx
console.log('name is ', name);
var name;
name = 'tom';
console.log('name is ', name);

// the result will be

// name is  undefined
// name is  tom
```

# Function Hoisting
Function Hoisting follows a similar path to Variable Hoisting. In Function Hoisting, JavaScript Execution Context's creation phase puts any function declaration existed in global scope directly into memory. And then, the functions of the memory can be executed in the execution phase when the function gets invoked. (Remember, function also creates its own Function Execution Context, the mechanism remains the same)

# Summary
Simply put, you should always declare the variable/function before you use it. Even though the keyword let & const are hoisted for declaration, but it’s not hoisted for initialization. That means, if you declare and assign value to the variable after you use it, the code won’t run even though let & const are hoisted because let and const declaration are not initialized to `undefined` like `var` do. Always define functions, variables, etc., before using them in code. Do not rely on Hoisting much. At the same time, it is important to understand the underlying concept of why certain things are behaving in specific ways.
